docs/guide/callbacks.rst

```diff
@@ -0,0 +1,326 @@
+
+
+.. code-block:: python
+
+    from stable_baselines.common.callbacks import BaseCallback
+
+
+    class CustomCallback(BaseCallback):
+        """
+        A custom callback that derives from `BaseCallback`.
+
+        :param verbose: (int) Verbosity level 0: not output 1: info 2: debug
+        """
+        def __init__(self, verbose=0):
+            super(CustomCallback, self).__init__(verbose)
+            # Those variables will be accessible in the callback
+            # (they are defined in the base class)
+            # The RL model
+            # self.model = None  # type: BaseRLModel
+            # An alias for self.model.get_env(), the environment used for training
+            # self.training_env = None  # type: Union[gym.Env, VecEnv, None]
+            # Number of time the callback was called
+            # self.n_calls = 0  # type: int
+            # self.num_timesteps = 0  # type: int
+            # local and global variables
+            # self.locals = None  # type: Dict[str, Any]
+            # self.globals = None  # type: Dict[str, Any]
+            # The logger object, used to report things in the terminal
+            # self.logger = None  # type: logger.Logger
+            # # Sometimes, for event callback, it is useful
+            # # to have access to the parent object
+            # self.parent = None  # type: Optional[BaseCallback]
+
+        def _on_training_start(self) -> None:
+            """
+            This method is called before the first rollout starts.
+            """
+            pass
+
+        def _on_rollout_start(self) -> None:
+            """
+            A rollout is the collection of environment interaction
+            using the current policy.
+            This event is triggered before collecting new samples.
+            """
+            pass
+
+        def _on_step(self) -> bool:
+            """
+            This method will be called by the model.
+            This is the equivalent to the callback function.
+            `locals()` and `globals()` are directly accessible as attributes.
+
+            :return: (bool) Return True for continuing training, False for stopping.
+            """
+            return True
+
+        def _on_rollout_end(self) -> None:
+            """
+            This event is triggered before updating the policy.
+            """
+            pass
+
+        def _on_training_end(self) -> None:
+            """
+            This event is triggered before exiting the `learn()` method.
+            """
+            pass
+
+
+.. note::
+  `self.num_timesteps` corresponds to the total number of steps taken in the environment, i.e., it is the number of environments multiplied by the number of time `env.step()` was called
+
+  You should know that `PPO1` and `TRPO` update `self.num_timesteps` after each rollout (and not each step) because they rely on MPI.
```

Comment:
Should highlight that same applies to PPO2/A2C/(some others?), too: in each call to _on_step the self.num_timesteps has increased by number of environments, not by one.

Reply:
Not really, what I meant is that `self.num_timesteps` is updated at each step (so + `n_envs`) for every algorithm except PPO1 and TRPO. So  `self.num_timesteps` in the callback is not up to date (only updated from time to time). I should clarify that.